src/ipc/ipc4/helper.c

```diff
@@ -767,6 +770,15 @@ int ipc4_chain_dma_state(struct comp_dev *dev, struct ipc4_chain_dma *cdma)
 		ret = comp_trigger(dev, COMP_TRIGGER_PAUSE);
 		if (ret < 0)
 			return ret;
+
+		list_for_item_safe(clist, _tmp, &ipc->comp_list) {
+			icd = container_of(clist, struct ipc_comp_dev, list);
+			if (icd->cd != dev)
+				continue;
+			list_item_del(&icd->list);
+			rfree(icd);
+			break;
+		}
 		comp_free(dev);
```

Comment:
I'd say it should be a one-liner
```
-	comp_free(dev);
+	ipc_comp_free(dev);
```
but one would need to check carefully that that function cannot do any harm. @plbossart could you maybe give it a quick test?

Reply:
update: the API seems to be asymmetric. There's `comp_free()` and there's `ipc_comp_free()` but I don't think anything like `comp_alloc()`, the closest I find is `ipc4_add_comp_dev()` and that's what chain DMA is using. So in the short run, I guess - whatever works, in the long run someone might want to clean that up a bit.

Reply:
Multiple issues:
a) int ipc_comp_free(struct ipc *ipc, uint32_t comp_id)
what is the comp_id?

b) 	if (!icd->cd->bsource_list.next || !icd->cd->bsink_list.next)
           return -EINVAL;
we don't have any such things defined for chain-dma, do we? My guess is that we would hit this error case.

I can give it a try once you guys help me with a)...

Reply:
Lets take this as a followup improvement, this is fixing a leak and unblocking tests.